docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,182 @@
+# Redefine Container Runtime Interface
+
+## Motivation
+
+Kubelet employs a declarative pod-level interface, which acts as the sole
+integration point for container runtimes (e.g., `docker` and `rkt`). The
+high-level, declarative interface has caused higher integration and maintenance
+cost, and also slowed down feature velocity for the following reasons.
+  1. **Not every container runtime supports the concept of pods natively**.
+     When integrating with Kubernetes, a significant amount of work needs to
+     go into implementing a shim of significant size to support all pod
+     features. This also adds maintenance overhead (e.g., `docker`).
+  2. **High-level interface discourage code sharing and reuse among runtimes**.
+     E.g, each runtime implements an all-encompassing `SyncPod()` function that
+     (re-)starts pods/containers and manage lifecycle hooks.
+  3. **Pod Spec is susceptible to inconsistent interpretations**. E.g., `rkt`
+     does not support container-level operations and assume immutable pods.
+  4. **Pod Spec is evolving rapidly**. New features are being added constantly.
+     Any pod-level change or addition requires changing of all container
+     runtime shims. E.g., init containers and volume containers.
+
+## Goals and Non-Goals
+
+The goals of defining the interface are to
+ - **improve extensibility**: Easier container runtime integration.
+ - **improve feature velocity**
+ - **improve code maintainability**
+
+The non-goals include
```

Comment:
We are going to version this API, but explicitly make this is non-goal for now? We are going to do this through Kubernetes v2 API anyway.